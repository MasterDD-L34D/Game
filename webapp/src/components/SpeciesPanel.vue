<template>
  <section class="species-panel" v-if="species">
    <SpeciesOverview :name="displayName" :summary="summary" :description="description">
      <template #actions>
        <SpeciesQuickActions @export="handleExport" @save="handleSave" />
      </template>
    </SpeciesOverview>

    <div class="species-panel__layout">
      <SpeciesBiology
        :core-traits="coreTraits"
        :derived-traits="derivedTraits"
        :adaptations="adaptations"
        :behaviour-tags="behaviourTags"
        :drives="behaviourDrives"
      >
        <template #filters>
          <TraitFilterPanel
            v-model="traitFilters"
            :core-options="availableTraits.core"
            :derived-options="availableTraits.derived"
          />
        </template>
      </SpeciesBiology>

      <aside class="species-panel__sidebar">
        <SpeciesStatistics :statistics="statistics" :synergy="formattedSynergy">
          <p v-if="metaInfo" class="species-panel__meta">{{ metaInfo }}</p>
        </SpeciesStatistics>
        <SpeciesRevisionTimeline :entries="timelineEntries" />
      </aside>
    </div>

    <SpeciesPreviewGrid :previews="previewCards" :loading="isPreviewLoading" :error="previewError">
      <template #filters>
        <button type="button" class="species-panel__refresh" @click="refreshPreviews" :disabled="isPreviewLoading">
          Aggiorna batch
        </button>
      </template>
    </SpeciesPreviewGrid>
  </section>
  <section v-else class="species-panel species-panel--empty">
    <p>Nessuna specie selezionata.</p>
  </section>
</template>

<script setup>
import { computed, ref, watch } from 'vue';

import SpeciesBiology from './species/SpeciesBiology.vue';
import SpeciesOverview from './species/SpeciesOverview.vue';
import SpeciesPreviewGrid from './species/SpeciesPreviewGrid.vue';
import SpeciesQuickActions from './species/SpeciesQuickActions.vue';
import SpeciesRevisionTimeline from './species/SpeciesRevisionTimeline.vue';
import SpeciesStatistics from './species/SpeciesStatistics.vue';
import TraitFilterPanel from './species/TraitFilterPanel.vue';
import { requestSpeciesPreviewBatch } from '../services/speciesPreviewService';

const props = defineProps({
  species: {
    type: Object,
    default: null,
  },
  validation: {
    type: Object,
    default: null,
  },
  meta: {
    type: Object,
    default: null,
  },
  previewBatch: {
    type: Array,
    default: () => [],
  },
  autoPreview: {
    type: Boolean,
    default: true,
  },
});

const emit = defineEmits(['export', 'save', 'preview-error']);

const summary = computed(() => props.species?.summary || null);
const description = computed(
  () => props.species?.description || props.species?.summary || ''
);
const traits = computed(() => props.species?.traits || {});
const coreTraits = computed(() => traits.value.core || props.species?.core_traits || []);
const derivedTraits = computed(
  () => traits.value.derived || props.species?.derived_traits || []
);
const morphology = computed(() => props.species?.morphology || {});
const adaptations = computed(() => morphology.value.adaptations || []);
const behaviour = computed(
  () => props.species?.behavior || props.species?.behavior_profile || {}
);
const behaviourTags = computed(
  () => behaviour.value.tags || behaviour.value.behaviourTags || []
);
const behaviourDrives = computed(() => behaviour.value.drives || []);
const statistics = computed(() =>
  props.species?.statistics || {
    threat_tier: props.species?.balance?.threat_tier,
    rarity: props.species?.rarity,
  }
);

const displayName = computed(
  () => props.species?.display_name || props.species?.id || 'Specie'
);

const metaInfo = computed(() => {
  const attempt = props.meta?.attempts;
  const fallback = props.meta?.fallback_used;
  if (attempt || fallback) {
    const segments = [];
    if (attempt) segments.push(`Tentativi: ${attempt}`);
    if (fallback != null) segments.push(`Fallback: ${fallback ? 'sì' : 'no'}`);
    return segments.join(' · ');
  }
  return '';
});

const formattedSynergy = computed(() => {
  const value = statistics.value?.synergy_score;
  if (typeof value === 'number' && Number.isFinite(value)) {
    return `${Math.round(value * 100)}%`;
  }
  return 'n/d';
});

const availableTraits = computed(() => ({
  core: Array.from(new Set(coreTraits.value || [])).filter(Boolean),
  derived: Array.from(new Set(derivedTraits.value || [])).filter(Boolean),
}));

const traitFilters = ref({ core: [], derived: [] });

const timelineEntries = computed(() => {
  const messages = Array.isArray(props.validation?.messages)
    ? props.validation.messages
    : [];
  return messages.map((message, index) => ({
    id: `${message.code || 'msg'}-${index}`,
    title: message.level === 'error' ? 'Errore' : message.level === 'warning' ? 'Avviso' : 'Info',
    message: message.message || message.code,
    code: message.code || 'n/d',
    level: message.level || 'info',
  }));
});

const previewCards = ref([]);
const isPreviewLoading = ref(false);
const previewError = ref('');

watch(
  () => props.previewBatch,
  (value) => {
    if (Array.isArray(value)) {
      previewCards.value = value;
    }
  },
  { immediate: true }
);

watch(
  () => props.species?.id,
  () => {
    traitFilters.value = {
      core: [...availableTraits.value.core],
      derived: [...availableTraits.value.derived],
    };
    previewCards.value = Array.isArray(props.previewBatch) ? props.previewBatch : [];
    previewError.value = '';
    if (props.autoPreview) {
      void refreshPreviews();
    }
  },
  { immediate: true }
);

watch(
  () => [
    traitFilters.value.core.slice().sort().join(','),
    traitFilters.value.derived.slice().sort().join(','),
    props.autoPreview,
    props.species?.id,
  ],
  async () => {
    if (props.autoPreview) {
      await refreshPreviews();
    }
  }
);

function buildPreviewRequests() {
  const coreSelection = traitFilters.value.core.length
    ? traitFilters.value.core
    : availableTraits.value.core;
  const derivedSelection = traitFilters.value.derived;
  const baseTraits = Array.from(new Set([...coreSelection, ...derivedSelection]));
  if (!baseTraits.length) {
    return [];
  }
  const biomeId = props.meta?.biome_id || morphology.value.environments?.[0] || null;
  const fallback = availableTraits.value.core;
  const combos = [baseTraits];
  for (const trait of derivedSelection) {
    combos.push(Array.from(new Set([...coreSelection, trait])));
  }
  const uniqueCombos = [];
  const seen = new Set();
  for (const combo of combos) {
    const signature = combo.slice().sort().join('|');
    if (!signature || seen.has(signature)) continue;
    seen.add(signature);
    uniqueCombos.push(combo);
  }
  return uniqueCombos.slice(0, 4).map((combo, index) => ({
    trait_ids: combo,
    biome_id: biomeId,
    seed: index,
    base_name: `${displayName.value} Preview ${index + 1}`,
    request_id: `${props.species?.id || 'preview'}-${index}`,
    fallback_trait_ids: fallback,
  }));
}

async function refreshPreviews() {
  if (!props.species) {
    previewCards.value = [];
    return;
  }
  const requests = buildPreviewRequests();
  if (!requests.length) {
    previewCards.value = [];
    return;
  }
  isPreviewLoading.value = true;
  previewError.value = '';
  try {
    const response = await requestSpeciesPreviewBatch(requests);
    previewCards.value = response.previews || [];
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    previewError.value = message;
    emit('preview-error', message);
  } finally {
    isPreviewLoading.value = false;
  }
}

function handleExport() {
  emit('export', props.species);
}

function handleSave() {
  emit('save', props.species);
}
</script>

<style scoped>
.species-panel {
  background: #0a0d12;
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: 12px;
  padding: 1.5rem;
  color: #f0f4ff;
  display: grid;
  gap: 1.5rem;
}

.species-panel__layout {
  display: grid;
  gap: 1.5rem;
  grid-template-columns: minmax(0, 1fr);
}

@media (min-width: 960px) {
  .species-panel__layout {
    grid-template-columns: minmax(0, 2fr) minmax(0, 1fr);
  }
}

.species-panel__sidebar {
  display: grid;
  gap: 1rem;
}

.species-panel__meta {
  margin: 0;
  font-size: 0.8rem;
  opacity: 0.75;
}

.species-panel__refresh {
  background: transparent;
  color: #f0f4ff;
  border: 1px solid rgba(240, 244, 255, 0.35);
  border-radius: 8px;
  padding: 0.4rem 0.9rem;
  cursor: pointer;
  font-size: 0.85rem;
  transition: background 0.15s ease;
}

.species-panel__refresh:hover:enabled {
  background: rgba(39, 121, 255, 0.18);
}

.species-panel__refresh:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.species-panel--empty {
  text-align: center;
  color: rgba(240, 244, 255, 0.6);
  border: 1px dashed rgba(255, 255, 255, 0.2);
  border-radius: 12px;
  padding: 2rem;
}
</style>
