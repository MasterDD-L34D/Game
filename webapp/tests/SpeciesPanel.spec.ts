import { flushPromises, mount } from '@vue/test-utils';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import SpeciesPanel from '../src/components/SpeciesPanel.vue';

vi.mock('../src/services/speciesPreviewService', () => ({
  requestSpeciesPreviewBatch: vi.fn(() => Promise.resolve({ previews: [] })),
}));

const { requestSpeciesPreviewBatch } = await import('../src/services/speciesPreviewService.js');

const BASE_SPECIES = {
  id: 'synthetic-1d37afd07a',
  display_name: 'Predatore / Coda a Frusta Cinetica',
  summary: 'Artigli a Sette Vie, Coda a Frusta Cinetica, Scheletro Idro-Regolante',
  description:
    'Sintesi genetica focalizzata su Artigli a Sette Vie e Coda a Frusta Cinetica con impronta morfologica defensive, locomotor, omeostatico, prehensile, structural; comportamento climber. Ottimizzata per biomi: caverna_risonante.',
  traits: {
    core: ['Artigli a Sette Vie', 'Coda a Frusta Cinetica', 'Scheletro Idro-Regolante'],
    derived: ['struttura_elastica_amorfa', 'sacche_galleggianti_ascensoriali'],
  },
  morphology: {
    adaptations: ['Dita lunghe', 'Precauzione: vulnerabile ai veleni'],
  },
  behavior_profile: {
    tags: ['climber'],
  },
  statistics: {
    threat_tier: 'T3',
    rarity: 'R2',
    energy_profile: 'medio',
    synergy_score: 0.42,
  },
};

const VALIDATION = {
  messages: [
    { level: 'info', code: 'species.schema_version.defaulted', message: 'schema_version mancante' },
    { level: 'warning', code: 'species.environment_affinity.missing', message: 'environment_affinity non presente' },
  ],
};

beforeEach(() => {
  vi.clearAllMocks();
});

describe('SpeciesPanel', () => {
  it('renders species narrative and mechanics', async () => {
    const wrapper = mount(SpeciesPanel, {
      props: { species: BASE_SPECIES, validation: VALIDATION, autoPreview: false },
    });
    expect(wrapper.text()).toContain('Predatore / Coda a Frusta Cinetica');
    expect(wrapper.text()).toContain('Tratti derivati');
    expect(wrapper.text()).toContain('Revisioni validate');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders placeholder when missing species', () => {
    const wrapper = mount(SpeciesPanel);
    expect(wrapper.text()).toContain('Nessuna specie selezionata');
  });

  it('emits quick action events', async () => {
    const wrapper = mount(SpeciesPanel, {
      props: { species: BASE_SPECIES, autoPreview: false },
    });
    await wrapper.find('button').trigger('click');
    await wrapper.findAll('button')[1].trigger('click');
    expect(wrapper.emitted('export')).toBeTruthy();
    expect(wrapper.emitted('save')).toBeTruthy();
  });

  it('requests preview batch when filters change', async () => {
    const wrapper = mount(SpeciesPanel, {
      props: { species: BASE_SPECIES },
    });
    await flushPromises();
    await vi.waitFor(() => {
      expect(requestSpeciesPreviewBatch).toHaveBeenCalled();
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    expect(checkbox.exists()).toBe(true);
    const callCount = requestSpeciesPreviewBatch.mock.calls.length;
    await checkbox.setValue(false);
    await flushPromises();
    await vi.waitFor(() => {
      expect(requestSpeciesPreviewBatch.mock.calls.length).toBeGreaterThan(callCount);
    });
  });
});
