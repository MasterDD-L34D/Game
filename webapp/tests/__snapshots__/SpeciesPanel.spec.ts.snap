// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpeciesPanel > renders species narrative and mechanics 1`] = `
"<section data-v-73eb6eb5="" class="species-panel">
  <section data-v-966c492d="" data-v-73eb6eb5="" class="nebula-shell">
    <div data-v-966c492d="" class="nebula-shell__frame">
      <header data-v-966c492d="" class="nebula-shell__header">
        <div data-v-966c492d="" class="nebula-shell__status-grid">
          <div data-v-966c492d="" class="nebula-shell__status"><span data-v-966c492d="" class="nebula-shell__status-led" data-tone="neutral"></span>
            <div data-v-966c492d="">
              <p data-v-966c492d="" class="nebula-shell__status-label">Rarit√†</p>
              <p data-v-966c492d="" class="nebula-shell__status-value">R2</p>
            </div>
          </div>
          <div data-v-966c492d="" class="nebula-shell__status"><span data-v-966c492d="" class="nebula-shell__status-led" data-tone="critical"></span>
            <div data-v-966c492d="">
              <p data-v-966c492d="" class="nebula-shell__status-label">Threat tier</p>
              <p data-v-966c492d="" class="nebula-shell__status-value">T3</p>
            </div>
          </div>
          <div data-v-966c492d="" class="nebula-shell__status"><span data-v-966c492d="" class="nebula-shell__status-led" data-tone="critical"></span>
            <div data-v-966c492d="">
              <p data-v-966c492d="" class="nebula-shell__status-label">Allineamento</p>
              <p data-v-966c492d="" class="nebula-shell__status-value">42%</p>
            </div>
          </div>
        </div>
        <div data-v-966c492d="" class="nebula-shell__actions">
          <div data-v-78ecd3ce="" data-v-73eb6eb5="" class="species-quick-actions"><button data-v-78ecd3ce="" type="button" class="species-quick-actions__button"> Esporta scheda </button><button data-v-78ecd3ce="" type="button" class="species-quick-actions__button species-quick-actions__button--secondary"> Salva nel pack </button></div>
        </div>
      </header>
      <div data-v-966c492d="" class="nebula-shell__cards">
        <article data-v-1960f20b="" data-v-73eb6eb5="" class="species-card">
          <header data-v-1960f20b="" class="species-card__header">
            <div data-v-1960f20b="">
              <p data-v-1960f20b="" class="species-card__kicker">Profilo non definito</p>
              <h3 data-v-1960f20b="" class="species-card__title">Predatore / Coda a Frusta Cinetica</h3>
            </div>
            <div data-v-1960f20b="" class="species-card__badges"><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="core"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ú¶</span><span data-v-48844140="" class="trait-chip__label">R2</span></span><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="hazard"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ö°</span><span data-v-48844140="" class="trait-chip__label">T3</span></span></div>
          </header>
          <p data-v-1960f20b="" class="species-card__synopsis">Artigli a Sette Vie, Coda a Frusta Cinetica, Scheletro Idro-Regolante</p>
          <section data-v-1960f20b="" class="species-card__traits" aria-label="Tratti principali">
            <div data-v-1960f20b="">
              <h4 data-v-1960f20b="">Core</h4>
              <div data-v-1960f20b="" class="species-card__trait-grid"><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="core"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ú∂</span><span data-v-48844140="" class="trait-chip__label">Artigli a Sette Vie</span></span><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="core"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ú∂</span><span data-v-48844140="" class="trait-chip__label">Coda a Frusta Cinetica</span></span><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="core"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ú∂</span><span data-v-48844140="" class="trait-chip__label">Scheletro Idro-Regolante</span></span></div>
            </div>
            <div data-v-1960f20b="">
              <h4 data-v-1960f20b="">Derivati</h4>
              <div data-v-1960f20b="" class="species-card__trait-grid"><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="derived"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ò≤</span><span data-v-48844140="" class="trait-chip__label">struttura_elastica_amorfa</span></span><span data-v-48844140="" data-v-1960f20b="" class="trait-chip" data-variant="derived"><span data-v-48844140="" class="trait-chip__icon" aria-hidden="true">‚ò≤</span><span data-v-48844140="" class="trait-chip__label">sacche_galleggianti_ascensoriali</span></span></div>
            </div>
            <!--v-if-->
          </section>
          <footer data-v-1960f20b="" class="species-card__footer">
            <div data-v-1960f20b="" class="species-card__info-block"><span data-v-1960f20b="" class="species-card__label">Energia</span><span data-v-1960f20b="">medio</span></div>
            <div data-v-1960f20b="" class="species-card__info-block"><span data-v-1960f20b="" class="species-card__label">Habitat</span><span data-v-1960f20b=""></span></div>
            <!--v-if-->
          </footer>
        </article>
      </div>
      <nav data-v-966c492d="" class="nebula-shell__tabs" aria-label="Navigazione schede"><button data-v-966c492d="" type="button" class="nebula-shell__tab"><span data-v-966c492d="" class="nebula-shell__tab-icon" aria-hidden="true">üß¨</span><span data-v-966c492d="">Scheda</span></button><button data-v-966c492d="" type="button" class="nebula-shell__tab"><span data-v-966c492d="" class="nebula-shell__tab-icon" aria-hidden="true">üåø</span><span data-v-966c492d="">Biologia</span></button><button data-v-966c492d="" type="button" class="nebula-shell__tab nebula-shell__tab--active"><span data-v-966c492d="" class="nebula-shell__tab-icon" aria-hidden="true">üì°</span><span data-v-966c492d="">Telemetry</span></button><button data-v-966c492d="" type="button" class="nebula-shell__tab"><span data-v-966c492d="" class="nebula-shell__tab-icon" aria-hidden="true">‚àû</span><span data-v-966c492d="">Sinergie</span></button></nav>
      <div data-v-966c492d="" class="nebula-shell__content">
        <div data-v-73eb6eb5="" class="species-panel__section species-panel__section--columns">
          <section data-v-e6469edf="" data-v-73eb6eb5="" class="species-statistics">
            <h3 data-v-e6469edf="">Statistiche</h3>
            <dl data-v-e6469edf="" class="species-statistics__grid">
              <div data-v-e6469edf="" class="species-statistics__stat">
                <dt data-v-e6469edf="">Minaccia</dt>
                <dd data-v-e6469edf="">T3</dd>
              </div>
              <div data-v-e6469edf="" class="species-statistics__stat">
                <dt data-v-e6469edf="">Rarit√†</dt>
                <dd data-v-e6469edf="">R2</dd>
              </div>
              <div data-v-e6469edf="" class="species-statistics__stat">
                <dt data-v-e6469edf="">Energia</dt>
                <dd data-v-e6469edf="">medio</dd>
              </div>
              <div data-v-e6469edf="" class="species-statistics__stat">
                <dt data-v-e6469edf="">Sinergia</dt>
                <dd data-v-e6469edf="">42%</dd>
              </div>
            </dl>
          </section>
          <section data-v-5b00ee54="" data-v-73eb6eb5="" class="species-timeline">
            <div data-v-5b00ee54="" class="species-timeline__header">
              <h3 data-v-5b00ee54="">Revisioni validate</h3>
              <div data-v-5b00ee54="" class="species-timeline__filters"><button data-v-5b00ee54="" type="button" class="species-timeline__filter species-timeline__filter--active"><span data-v-5b00ee54="" class="species-timeline__filter-icon">‚ú¶</span><span data-v-5b00ee54="" class="species-timeline__filter-label">Tutte</span><span data-v-5b00ee54="" class="species-timeline__filter-count">2</span></button><button data-v-5b00ee54="" type="button" class="species-timeline__filter"><span data-v-5b00ee54="" class="species-timeline__filter-icon">‚úß</span><span data-v-5b00ee54="" class="species-timeline__filter-label">Info</span><span data-v-5b00ee54="" class="species-timeline__filter-count">1</span></button><button data-v-5b00ee54="" type="button" class="species-timeline__filter"><span data-v-5b00ee54="" class="species-timeline__filter-icon">‚ö†Ô∏è</span><span data-v-5b00ee54="" class="species-timeline__filter-label">Avvisi</span><span data-v-5b00ee54="" class="species-timeline__filter-count">1</span></button><button data-v-5b00ee54="" type="button" class="species-timeline__filter"><span data-v-5b00ee54="" class="species-timeline__filter-icon">‚ò†Ô∏è</span><span data-v-5b00ee54="" class="species-timeline__filter-label">Critici</span><span data-v-5b00ee54="" class="species-timeline__filter-count">0</span></button><button data-v-5b00ee54="" type="button" class="species-timeline__filter"><span data-v-5b00ee54="" class="species-timeline__filter-icon">‚ú®</span><span data-v-5b00ee54="" class="species-timeline__filter-label">OK</span><span data-v-5b00ee54="" class="species-timeline__filter-count">0</span></button></div>
            </div>
            <ol data-v-5b00ee54="" class="species-timeline__list">
              <li data-v-5b00ee54="" class="species-timeline__entry" data-tone="info">
                <div data-v-5b00ee54="" class="species-timeline__entry-header"><span data-v-5b00ee54="" class="species-timeline__badge" data-tone="info"><span data-v-5b00ee54="" class="species-timeline__badge-icon">‚úß</span><span data-v-5b00ee54="" class="species-timeline__badge-text">Info</span></span>
                  <div data-v-5b00ee54="" class="species-timeline__entry-meta"><strong data-v-5b00ee54="">Info</strong><span data-v-5b00ee54="" class="species-timeline__code">species.schema_version.defaulted</span></div>
                </div>
                <p data-v-5b00ee54="" class="species-timeline__message">schema_version mancante</p>
              </li>
              <li data-v-5b00ee54="" class="species-timeline__entry" data-tone="warning">
                <div data-v-5b00ee54="" class="species-timeline__entry-header"><span data-v-5b00ee54="" class="species-timeline__badge" data-tone="warning"><span data-v-5b00ee54="" class="species-timeline__badge-icon">‚ö†Ô∏è</span><span data-v-5b00ee54="" class="species-timeline__badge-text">Avviso</span></span>
                  <div data-v-5b00ee54="" class="species-timeline__entry-meta"><strong data-v-5b00ee54="">Avviso</strong><span data-v-5b00ee54="" class="species-timeline__code">species.environment_affinity.missing</span></div>
                </div>
                <p data-v-5b00ee54="" class="species-timeline__message">environment_affinity non presente</p>
              </li>
            </ol>
          </section>
          <!--v-if-->
          <!--v-if-->
        </div>
      </div>
      <footer data-v-966c492d="" class="nebula-shell__footer"></footer>
    </div>
  </section>
</section>"
`;
